generator/transit_generator.cpp

```diff
@@ -370,14 +444,140 @@ bool GraphData::IsSorted() const
+  // Set with stop ids with stops which are inside |borders|.
+  set<StopId> stopIdInside;
+  for (auto const & s : m_stops)
+  {
+    if (m2::RegionsContain(borders, s.GetPoint()))
+      stopIdInside.insert(s.GetId());
+  }
+
+  set<StopId> hasNeighborInside;
+  for (auto const & e : m_edges)
+  {
+    if (stopIdInside.count(e.GetStop1Id()) != 0 && stopIdInside.count(e.GetStop2Id()) == 0)
+      hasNeighborInside.insert(e.GetStop2Id());
+    if (stopIdInside.count(e.GetStop2Id()) != 0 && stopIdInside.count(e.GetStop1Id()) == 0)
+      hasNeighborInside.insert(e.GetStop1Id());
+  }
+
+  // Filling |lines| with stops inside |borders|.
+  vector<Line> lines;
+  for (auto const & l : m_lines)
+  {
+    Ranges const & stopIds = l.GetStopIds();
+    CHECK_EQUAL(stopIds.size(), 1, ());
+    // Note. |stopIdsToFill| will be filled with continuous sequences of stop ids.
+    // In most cases only one sequence of stop ids should be place to |stopIdsToFill|.
+    // But if a line is split by |borders| several times then several
+    // continuous sequences of stop ids will be place to |stopIdsToFill|.
+    // The loop below goes through all the stop ids belong the line |l| and keep in |stopIdsToFill|
+    // continuous sequences of stop ids which is inside |borders|.
+    Ranges stopIdsToFill;
+    bool stopWasOutsideBorders = true;
+    for (auto const stopId : stopIds[0])
+    {
+      if (stopIdInside.count(stopId) != 0 || hasNeighborInside.count(stopId) != 0)
+      {
+        if (stopWasOutsideBorders)
+        {
+          stopIdsToFill.emplace_back(); // Push an empty vector.
+          stopWasOutsideBorders = false;
+        }
+        stopIdsToFill.back().push_back(stopId);
+      }
+      else
+      {
+        stopWasOutsideBorders = true;
+      }
+    }
+
+    if (!stopIdsToFill.empty())
+    {
+      lines.emplace_back(l.GetId(), l.GetNumber(), l.GetTitle(), l.GetType(), l.GetColor(),
+                         l.GetNetworkId(), stopIdsToFill, l.GetInterval());
+    }
+  }
+  m_lines = move(lines);
+}
+
+void GraphData::ClipStops()
+{
+  CHECK(is_sorted(m_stops.cbegin(), m_stops.cend()), ());
+  set<StopId> stopIds;
+  for (auto const & l : m_lines)
+  {
+    for (auto const & range : l.GetStopIds())
+      stopIds.insert(range.cbegin(), range.cend());
+  }
+
+  UpdateItems(stopIds, m_stops);
+}
+
+void GraphData::ClipNetworks()
+{
+  CHECK(is_sorted(m_networks.cbegin(), m_networks.cend()), ());
+  set<NetworkId> networkIds;
+  for (auto const & l : m_lines)
+    networkIds.insert(l.GetNetworkId());
+
+  UpdateItems(networkIds, m_networks);
+}
+
+void GraphData::ClipGates()
+{
+  ClipItemsByStops(m_stops, m_gates);
+}
+
+void GraphData::ClipTransfer()
+{
+  ClipItemsByStops(m_stops, m_transfers);
+}
+
+void GraphData::ClipEdges()
+{
+  CHECK(is_sorted(m_stops.cbegin(), m_stops.cend()), ());
+
+  vector<Edge> edges;
+  for (auto const & edge : m_edges)
+  {
+    if (HasStop(m_stops, edge.GetStop1Id()) && HasStop(m_stops, edge.GetStop2Id()))
```

Comment:
Is it correct? Do we need to replace ```&&``` by ```II```?

Reply:
we have stops which are inside and stops which are directly connected to stops inside in m_stops, so I think it's correct

Reply:
Yes, an edge belongs to an mwm if the both its stops belong to the mwm.